src/mlpack/methods/logistic_regression/logistic_regression_function_impl.hpp

```diff
@@ -235,6 +225,78 @@ void LogisticRegressionFunction<MatType>::PartialGradient(
   }
 }
 
+template<typename MatType>
+template<typename GradType>
+double LogisticRegressionFunction<MatType>::EvaluateWithGradient(
+    const arma::mat& parameters,
+    GradType& gradient) const
+{
+  // Regularization term.
+  arma::mat regularization = lambda *
+      parameters.tail_cols(parameters.n_elem - 1);
+
+  const double objectiveRegularization = lambda / 2.0 *
+      arma::dot(parameters.tail_cols(parameters.n_elem - 1),
+                parameters.tail_cols(parameters.n_elem - 1));
+
+  // Calculate the sigmoid function values.
+  const arma::rowvec sigmoids = 1.0 / (1.0 + arma::exp(-(parameters(0, 0) +
+      parameters.tail_cols(parameters.n_elem - 1) * predictors)));
+
+  gradient.set_size(arma::size(parameters));
+  gradient[0] = -arma::accu(responses - sigmoids);
+  gradient.tail_cols(parameters.n_elem - 1) = (sigmoids - responses) *
+      predictors.t() + regularization;
+
+  // Now compute the objective function using the sigmoids.
+  double result = arma::accu(arma::log(1.0 -
+      arma::conv_to<arma::rowvec>::from(responses) + sigmoids %
+      (2 * arma::conv_to<arma::rowvec>::from(responses) - 1.0)));
+
+  // Invert the result, because it's a minimization.
+  return -result + objectiveRegularization;
+}
+
+template<typename MatType>
+template<typename GradType>
+double LogisticRegressionFunction<MatType>::EvaluateWithGradient(
+    const arma::mat& parameters,
+    const size_t begin,
+    GradType& gradient,
+    const size_t batchSize) const
+{
+  // Regularization term.
+  arma::mat regularization =
+      lambda * parameters.tail_cols(parameters.n_elem - 1) / predictors.n_cols *
+      batchSize;
+
+  const double objectiveRegularization = lambda *
+      (batchSize / (2.0 * predictors.n_cols)) *
+      arma::dot(parameters.tail_cols(parameters.n_elem - 1),
+                parameters.tail_cols(parameters.n_elem - 1));
+
+  // Calculate the sigmoid function values.
+  const arma::rowvec sigmoids = 1.0 / (1.0 + arma::exp(-(parameters(0, 0) +
+      parameters.tail_cols(parameters.n_elem - 1) *
+      predictors.cols(begin, begin + batchSize - 1))));
+
+  gradient.set_size(parameters.n_rows, parameters.n_cols);
+  gradient[0] = -arma::accu(responses.subvec(begin, begin + batchSize - 1) -
+      sigmoids);
+  gradient.tail_cols(parameters.n_elem - 1) = (sigmoids -
+      responses.subvec(begin, begin + batchSize - 1)) *
+      predictors.cols(begin, begin + batchSize - 1).t() + regularization;
+
+  // Now compute the objective function using the sigmoids.
+  arma::rowvec respD = arma::conv_to<arma::rowvec>::from(responses.subvec(begin,
+      begin + batchSize - 1));
+  const double result = arma::accu(arma::log(1.0 - respD + sigmoids %
+      (2 * respD - 1.0)));
+
+  // Invert the result, because it's a minimization.
+  return -result + objectiveRegularization;
```

Comment:
Should we write ``objectiveRegularization - result``?

Reply:
Good point, that is easier to read.  Fixed.